proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,165 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+Exception Types
+
+```
+exntype ::= [t*] -> []
+```
+
+### Instructions
+
+```
+instr ::= ... | throw x | rethrow l
+        | try bt instr* (catch x instr*)+ end
+        | try bt instr* (catch x instr*)* catch_all instr* end
+        | try bt instr* delegate l
+        | try bt instr* unwind instr* end
+```
+
+### Modules
+
+Exceptions (definitions)
+
+```
+exn ::= export* exception exntype  | export* exception exntype import
+```
+
+Modules
+
+
+```
+mod ::= module ... exn*
+```
+
+## Validation (Typing)
+
+To verify that a `try...delegate l` instruction refers to a label introduced by a try block (I'll call this a try-label), introduce a type attribute to labels in the validation context, which type attribute is set to `try` when the label is a try-label. The original notation `label [t*]` will then be a shortcut for `label {result [t*], type ε}`.
+
+Moreover, to verify that the `rethrow l` instruction refers to a label introduced by a catch block (call this a catch-label), we allow this type attribute of labels in the validation context to be set to `catch` when the label is a catch label. This addition is reflected in the execution rules, by the administrative instruction `caught` which introduces a label around the instructions of the catching try-block's catch instructions.
```

Comment:
So the intention here is, as in the case of `delegate`, `rethrow` can target any label including labels for blocks, loops, and try-delegates, but they are just validation failures, right?

The concept of catch label is a little new. Can we consider it in the same as try label? Then we have this problem:
```wasm
try $label0
  rethrow $label0  ;; cannot be done, because it's not within catch below
catch
end
```

But `delegate` has a similar problem anyway if it takes a try block:
```wasm
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

The problem here is, in reality both instructions target `catch`, not `try`. So what do you think? Should we introduce a concept for catch label for both of them? Then with respect to normal branches, should we say catch label does not exist or the same as its corresponding try label?